mapper/dm_agent/src/batcher.rs

```diff
@@ -144,14 +145,24 @@ impl MessageBatcher {
         messages: &mut dyn MqttMessageStream,
         mut timeout: Interval,
     ) -> Result<MeasurementGrouper, DeviceMonitorError> {
-        let mut message_batch = MessageBatch::start_batch(first_message)?;
         timeout.tick().await; // The first tick starts the timeout window
+        let collectd_message = CollectdMessage::parse_from(&first_message)?;
```

Comment:
What if the very first message is ill-formed? The timeout window is then erroneously started.

To my understanding, we need to swap the two lines:

```
let collectd_message = CollectdMessage::parse_from(&first_message)?;
timeout.tick().await; // The first tick starts the timeout window
let mut message_batch = MessageBatch::start_batch(collectd_message)?;
```

Reply:
It won't result in an empty batch as the parsing will fail here and the resultant error will propogate to the caller, where it's logged. And the calling method will start a fresh batch with a fresh timeout. But yeah, the original `tick` becomes pointless as that batch doesn't even start. So I can swap the calls.

Reply:
Looking closer, I think the code would be easier to understand if the timeout argument of the `build_message_batch_with_timeout` function would be a `Duration` not an `Interval`.

Indeed, the current code is correct, but fragile.
* The first `timeout.tick().await` is not waiting for some time to elapse because this is the very *first* call of the `tick` method on a *fresh* `Interval` that is created on the call site. 
* If someone tries to re-use the `Interval` struct across the calls to `build_message_batch_with_timeout`, then the behavior will be weird in an hardly detectable manner.

So I suggest:
* to pass a `Duration` rather than an `Interval` to `build_message_batch_with_timeout` .
* to consider replacing the `timeout::tick()` by a call to `tokio::time::sleep()`.

Reply:
I used `Interval` instead of `Duration` with the hope of better controlling when the tick happens from tests. My goal was to pass a mocked interval somehow and externally control when the `tick` completes. Since I couldn't quite achieve that, now I can use a `Duration` instead.

Reply:
Using a `Duration` will really improve the function. For `sleep` vs `tick`, I let you see what is the most convenient.